### PR TITLE
Fix EnC and EE for ref features

### LINKED ISSUE #19274: Enable EnC and EE scenarios for generating metadata for IsReadOnlyAttribute
Work is done for compilation in #18715. We need to make sure it is true for both EnC and EE as well:

1. Generating IsReadOnly attribute if needed.
2. Evaluating ref-readonly-ness from source if necessary.
3. Overload resolution on such parameters.
4. Using attributes from referenced assemblies vs generating a new one.

Related to #17287 and #19216

### PR DESCRIPTION
**Pull request by OmarTawfik**
This PR Fixes #19274, and does the following:

* Disables EnC for ref structs.
* Disables EnC for readonly structs.
* Disables EnC for ref readonly parameters and return types.
* Enables Expression Evaluation for lambdas containing ref readonly signatures.

cc @ivanbasov @VSadov @cston @dotnet/roslyn-compiler

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs [hunk 1]
```diff
@@ -152,6 +152,12 @@ private void SetGlobalErrorIfTrue(bool arg)
                     methodCompiler.CompileSynthesizedMethods(additionalTypes, diagnostics);
                 }
 
+                var embeddedTypes = moduleBeingBuiltOpt.GetEmbeddedTypes(diagnostics);
+                if (!embeddedTypes.IsEmpty)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs]: The `if` seems unnecessary.

File: src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/ExpressionCompilerTests.cs [hunk 2]
```diff
@@ -6309,5 +6309,78 @@ .maxstack  1
...
+        public void RefReadOnlyLambdasEvaluationWillSynthesizeRequiredAttributes_Parameters()
+        {
+            var reference = CreateStandardCompilation(@"
+public delegate void D(ref readonly int p);");
+
+            CompileAndVerify(reference, symbolValidator: module =>
+            {
+                Assert.NotNull(module.ContainingAssembly.GetTypeByMetadataName(AttributeDescription.CodeAnalysisEmbeddedAttribute.FullName));
+                Assert.NotNull(module.ContainingAssembly.GetTypeByMetadataName(AttributeDescription.IsReadOnlyAttribute.FullName));
+            });
```
Comment (cston) [inline src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/ExpressionCompilerTests.cs]: Consider extracting the `module => { ... }` lambda to a helper method.

Comment (OmarTawfik) [reply src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/ExpressionCompilerTests.cs]: They're doing the opposite. First we assert it is there, and then assert it is not there.

File: src/EditorFeatures/CSharpTest/EditAndContinue/TopLevelEditingTests.cs [hunk 3]
```diff
@@ -676,6 +676,66 @@ public interface I
...
+            string src1 = "";
+            string src2 = "ref struct X { }";
+
+            var edits = GetTopEdits(src1, src2, CSharpParseOptions.Default.WithLanguageVersion(LanguageVersion.Latest));
+
+            edits.VerifyEdits(
+                "Insert [ref struct X { }]@0");
+
+            edits.VerifyRudeDiagnostics(
+                Diagnostic(RudeEditKind.RefStruct, "ref struct X", "struct"));
```
Comment (ivanbasov) [inline src/EditorFeatures/CSharpTest/EditAndContinue/TopLevelEditingTests.cs]: >struct [](start = 68, length = 6)

reference to feature resources? here and below

File: src/EditorFeatures/CSharpTest/EditAndContinue/StatementEditingTests.cs [hunk 4]
```diff
@@ -6659,6 +6659,81 @@ static void Main(int X)
...
+            var src1 = @"class Test { void M() { } }";
+            var src2 = @"class Test { void M() { void local(ref readonly int b) { throw null; } } }";
+
+            var edits = GetTopEdits(src1, src2, CSharpParseOptions.Default.WithLanguageVersion(LanguageVersion.Latest));
+
+            edits.VerifyEdits(
+                "Update [void M() { }]@13 -> [void M() { void local(ref readonly int b) { throw null; } }]@13");
+
+            edits.VerifyRudeDiagnostics(
+                Diagnostic(RudeEditKind.ReadOnlyReferences, "ref readonly int b", "parameter"));
```
Comment (ivanbasov) [inline src/EditorFeatures/CSharpTest/EditAndContinue/StatementEditingTests.cs]: >parameter [](start = 83, length = 9)

reference to feature resources here and below

**Comment by OmarTawfik** [comment]
@VSadov can you take a look? `ref structs` and `readonly structs` are included here.

File: src/EditorFeatures/CSharpTest/EditAndContinue/TopLevelEditingTests.cs (hunk 3 above)
Comment (OmarTawfik) [reply src/EditorFeatures/CSharpTest/EditAndContinue/TopLevelEditingTests.cs]: Replacing with `SyntaxFacts.GetText(SyntaxKind.StructKeyword)`

---
In reply to: [139556357](https://github.com/dotnet/roslyn/pull/22177#discussion_r139556357) [](ancestors = 139556357)

**Pull request merged**